presto-spi/src/main/java/com/facebook/presto/spi/type/Type.java

```diff
@@ -141,4 +142,9 @@
      * Compare the values in the specified block at the specified positions equal.
      */
     int compareTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition);
+
+    /**
+     * Get the encoding for this type.
+     */
+    BlockEncoding getEncoding();
```

Comment:
I think the correct thing to do is to use a `BlockEncodingSerde` in all the places that call this method. However, that's difficult because many places that use this to decode a `Slice` into a `Block` are scalar functions.

Another option is to change the native container type (the class returned from `getJavaType()`) of `ArrayType` and `MapType` to be `Block.class`. This has the additional benefit that we don't need to serialize and deserialize the arrays/maps into `Slice`s, so it should be much faster, it will also be more work since there's a lot of places in the code base with switch statements that assume there are only 4 native container types.

@dain thoughts on whether we want to do this as an intermediate step, or just go straight to using `Block` as the native container type for array/map?